### PR TITLE
fwupd: 1.0.1 → 1.0.4

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
Update and clean-up.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by orivej — APPROVED** [review]

**Comment by jtojnar** [comment]
Maybe we can wait for the [pull request](https://github.com/rhboot/efivar/pull/94) to be merged so we do not include the patch in repo?

**Comment by disassembler** [comment]
@GrahamcOfBorg build fwupd

**Comment by disassembler** [comment]
looks good! I don't see any problems with merging with the patch, but if you want to wait, that's fine too.

**Review by disassembler — APPROVED** [review]

**Comment by jtojnar** [comment]
Waiting few days is probably fine.

Regarding the aarch64 failure, we could remove `libsmbios` dependency on that platform. Unless `smbios` is supported on ARM and the limitation is only artificial.

**Comment by grahamc** [comment]
> Originally designed for Intel® processor architecture systems, SMBIOS now includes support for IA-32 (x86), x64 (x86-64, Intel64, AMD64, EM64T), Intel® Itanium® architecture, 32-bit ARM (Aarch32) and 64-bit ARM (Aarch64).

however our libsmbios does not live up to this claim:

```
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I./out/include -I./out/include -I./src/include -DLIBSMBIOS_LOCALEDIR=\"/nix/store/fvdfb1cy430j35n0vpb5c6n7a48sf2bi-libsmbios-2.3.3/share/locale\" -I./src/libsmbios_c/common -Wall -fPIC -fvisibility=hidden -g -O2 -DLIBSMBIOS_C_ALL_DYN_LINK -c src/libsmbios_c/token/token_da.c  -fPIC -DPIC -o src/libsmbios_c/token/.libs/out_libsmbios_c_la-token_da.o
src/libsmbios_c/cmos/cmos_linux.c:25:20: fatal error: sys/io.h: No such file or directory
 #include <sys/io.h>
                    ^
compilation terminated.
make[1]: *** [Makefile:2599: src/libsmbios_c/cmos/out_libsmbios_c_la-cmos_linux.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
src/libsmbios_c/smi/smi_linux.c: In function 'LINUX_dell_smi_obj_execute':
src/libsmbios_c/smi/smi_linux.c:308:9: warning: ignoring return value of 'strerror_r', declared with attribute warn_unused_result [-Wunused-result]
         strerror_r(errno, this->errstring + curstrsize, ERROR_BUFSIZE - curstrsize - 1);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: Leaving directory '/build/source'
make: *** [Makefile:3313: all-recursive] Error 1
builder for '/nix/store/15wkv5b5w3rfvcc8hxmvvs7i4r0d092v-libsmbios-2.3.3.drv' failed with exit code 2
cannot build derivation '/nix/store/grsmim0p1sp71rm5hj5a68np3znfkkyg-fwupd-1.0.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/grsmim0p1sp71rm5hj5a68np3znfkkyg-fwupd-1.0.2.drv' failed
```

I'm not sure we strictly need to handle the aarch64 failure at this time.

**Comment by dtzWill** [comment]
beep boop, ping! 1.0.4 is now out, FWIW!

**Comment by dtzWill** [comment]
This seems to do the trick:
```diff
From 6fb73e976372fbebf00e3d27e465dfa06121c3ee Mon Sep 17 00:00:00 2001
From: Will Dietz <w@wdtz.org>
Date: Fri, 9 Feb 2018 14:37:23 -0600
Subject: [PATCH] fwupd: 1.0.3 -> 1.0.4

---
 pkgs/os-specific/linux/firmware/fwupd/default.nix | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/pkgs/os-specific/linux/firmware/fwupd/default.nix b/pkgs/os-specific/linux/firmware/fwupd/default.nix
index cf2216d3916..a67a052e5fe 100644
--- a/pkgs/os-specific/linux/firmware/fwupd/default.nix
+++ b/pkgs/os-specific/linux/firmware/fwupd/default.nix
@@ -1,18 +1,16 @@
-{ stdenv, fetchFromGitHub, gtk_doc, pkgconfig, gobjectIntrospection, intltool
+{ stdenv, fetchurl, gtk_doc, pkgconfig, gobjectIntrospection, intltool
 , libgudev, polkit, appstream-glib, gusb, sqlite, libarchive, glib_networking
 , libsoup, docbook2x, gpgme, libxslt, libelf, libsmbios, efivar, glibcLocales
 , fwupdate, libyaml, valgrind, meson, libuuid, pygobject3, colord
-, pillow, ninja, gcab, gnutls, python3Packages, wrapGAppsHook
+, pillow, ninja, gcab, gnutls, python3Packages, wrapGAppsHook, json_glib
 }:
 let
-  version = "1.0.3";
+  version = "1.0.4";
 in stdenv.mkDerivation {
   name = "fwupd-${version}";
-  src = fetchFromGitHub {
-    owner = "hughsie";
-    repo = "fwupd";
-    rev = version;
-    sha256 = "0vg9f1gzcvfd7jnfhvv0ylf04fy7cw31ykrim7qik6gbkhr4gr65";
+  src = fetchurl {
+    url = "https://people.freedesktop.org/~hughsient/releases/fwupd-${version}.tar.xz";
+    sha256 = "1n4d6fw3ffg051072hbxn106s52x2wlh5dh2kxwdfjsb5kh03ra3";
   };
 
   nativeBuildInputs = [
@@ -21,7 +19,7 @@ in stdenv.mkDerivation {
   ];
   buildInputs = [
     polkit appstream-glib gusb sqlite libarchive libsoup libelf libsmbios fwupdate libyaml
-    libgudev colord gpgme libuuid pillow gnutls glib_networking efivar
+    libgudev colord gpgme libuuid pillow gnutls glib_networking efivar json_glib
   ];
 
   LC_ALL = "en_US.UTF-8"; # For po/make-images
-- 
2.16.1
```

**Comment by dtzWill** [comment]
Hmm, for some reason using this (1.0.4) as well at least on master (1.0.1) doesn't seem to find updates.

Logs show messages like:

```
/var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz file: cannot process file of type text/plain
```

Which is strange, considering:
```shell
$ file /var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz
/var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz: gzip compressed data, from Unix
```

------
Oh! Searching on upstream's github it appears you've discussed this and we need to update gcab, maybe?

https://github.com/hughsie/fwupd/issues/318#issuecomment-346962118

**Comment by dtzWill** [comment]
https://github.com/dtzWill/nixpkgs/tree/experimental/fwupd bumps our gcab and appstream-glib (with no testing of anything other than blitzing a path to fwupd :innocent), if that's helpful.

Unfortunately still seeing the problem re:unable to parse text/plain.  Dunno.

**Comment by jtojnar** [comment]
> Oh! Searching on upstream's github it appears you've discussed this and we need to update gcab, maybe?

The issue I was reporting was caused by an broken error handler, it has nothing to do with gcab and it was fixed in https://github.com/hughsie/fwupd/commit/5f733f25817356ee1368e2e1407f88b36e8d785a – the “cannot process file of type text/plain” only triggered the bug.

I noticed your issue in [1.0.1](https://github.com/NixOS/nixpkgs/pull/30252#issuecomment-342490756) but since I only needed fwupd as a build dependency for gnome-software, I did not pay it much heed. It should probably be reported upstream.

**Comment by dtzWill** [comment]
Looks like we need `shared_mime_info`, something like this in preFixup: `gappsWrapperArgs+=(--prefix XDG_DATA_DIRS : "${shared_mime_info}/share")`.

That seems to fix problems parsing the remotes' XML! Woo. But nothing seems to happen when I request the updates are performed/applied.

**Comment by jtojnar** [comment]
@dtzWill I merged your `gcab` update and `appstream-glib` update from my other branch.

And efivar was updated just now so I think this is ready to merge.

File: nixos/release.nix [hunk 1]
```diff
@@ -258,6 +258,7 @@ in rec {
   tests.firefox = callTest tests/firefox.nix {};
   tests.firewall = callTest tests/firewall.nix {};
   tests.fleet = callTestOnTheseSystems ["x86_64-linux"] tests/fleet.nix {};
+  tests.fwupd = callTest tests/fwupd.nix {};
```
Comment (vcunat) [inline nixos/release.nix]: This is a problem on Hydra.  It causes build during evaluation, most likely due to the code around `listFiles`.  It's probably the cause why it doesn't have resources to evaluate trunk-combined ATM.  I disabled the test for now.

Comment (vcunat) [reply nixos/release.nix]: There's an issue already :-)  https://github.com/NixOS/nixpkgs/issues/32142

Comment (jtojnar) [reply nixos/release.nix]: Right, sorry about that. Still the same old https://github.com/NixOS/nixpkgs/issues/32142, still not sure how to best fix it.

**Pull request merged**